src/ProcessHosts/Elastic.ProcessHosts/Process/ProcessBase.cs

```diff
@@ -56,40 +65,52 @@ public abstract class ProcessBase : IDisposable
 		public virtual void Start()
 		{
 			this.Stop();
+			this.RunningState = RunningState.Starting;
 			this.StartedHandle = new ManualResetEvent(false);
 
 			var bin = this.ProcessExe;
 			var args = this.Arguments;
 
-			var observable = this._process.Start(bin, args).Publish();
-			if (this._process.UserInteractive)
+			try
 			{
-				//subscribe to all messages and write them to console
-				this.Disposables.Add(observable.Subscribe(this._handler.Write, delegate { }, delegate { }));
+				var observable = this._process.Start(bin, args).Publish();
+				if (this._process.UserInteractive)
+				{
+					//subscribe to all messages and write them to console
+					this.Disposables.Add(observable.Subscribe(this._handler.Write, delegate { }, delegate { }));
+				}
+
+				//subscribe as long we are not in started state and attempt to read console
+				//out for this confirmation
+				this.Disposables.Add(observable
+					.TakeWhile(c => !this.Started)
+					.Subscribe(this.Handle, delegate { }, delegate { })
+				);
+				this.Disposables.Add(observable.Subscribe(delegate { }, HandleException, HandleCompleted));
+				this.Disposables.Add(observable.Connect());
+
+				var timeout = this._process.WaitForStarted;
+				//we wait for 1 minute to attempt a clean start (meaning when the service is started elasticsearch is started)
+				//this is a best effort, elasticsearch could take longer to start if it needs to relocate shards for instance
+				this.StartedHandle.WaitOne(timeout, true);
+
+				if (this.RunningState != RunningState.ConfirmedStarted) this.RunningState = RunningState.AssumedStarted;
```

Comment:
if the process is in `Stopping` or `Stopped` state, which may signal that there is an issue in starting Elasticsearch e.g. invalid configuration setting, should the state be `AssumedStarted`, or is this grounds to fail the installation? I totally see the merits to taking the assume started approach, which implies a philosophy of best effort to starting the service.

Waiting on the service to start could also be controlled with the [`StartOn` property](https://github.com/elastic/windows-installers/blob/c00a0d648cb40cdb37f1fa4532946181dbbf37fa/src/Installer/Elastic.Installer.Msi/Elasticsearch/Elasticsearch.cs#L85) on `ServiceInstall` entry, if we didn't want the installation process to wait at all for startup

Reply:
@forloop not sure if I can unpack your first paragraph. Why do you think it should be `AssumedStarted` in this case?

I looked at `StartOn` but I think the current approach is best, folks can now control starting the server on install and upgrade in the silent install and UI.

Reply:
> Why do you think it should be `AssumedStarted` in this case?

It will end up being `AssumeStarted` in this case, even though the service was started by the installer, but stopped, and the `RunningState` at the point of the check for `if (this.RunningState != RunningState.ConfirmedStarted)` is either `Stopping` or `Stopped`.

As an example, if you build 6.1.1 and 6.1.2 installers from this branch then

1. install 6.1.1 with X-Pack, and start service as part of install
2. upgrade 6.1.2 _without_ X-Pack, and start service as part of install

then the service is started as part of upgrade, but the process stops because the Elasticsearch keystore contains a `bootstrap.password` key, which is an unknown key when Elasticsearch starts, and therefore fails the bootstrap checks, before the started message is logged 

```
[2018-02-08T10:57:52,578][INFO ][o.e.n.Node               ] [LENOVO] initializing ...
[2018-02-08T10:57:52,648][INFO ][o.e.e.NodeEnvironment    ] [LENOVO] using [1] data paths, mounts [[(C:)]], net usable_space [119.6gb], net total_space [465.2gb], types [NTFS]
[2018-02-08T10:57:52,649][INFO ][o.e.e.NodeEnvironment    ] [LENOVO] heap size [1.9gb], compressed ordinary object pointers [true]
[2018-02-08T10:57:52,663][INFO ][o.e.n.Node               ] [LENOVO] node name [LENOVO], node ID [bVJxF-fnTR22qCbsbDyJzw]
[2018-02-08T10:57:52,663][INFO ][o.e.n.Node               ] [LENOVO] version[6.1.2], pid[8884], build[5b1fea5/2018-01-10T02:35:59.208Z], OS[Windows 10/10.0/amd64], JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/1.8.0_144/25.144-b01]
[2018-02-08T10:57:52,663][INFO ][o.e.n.Node               ] [LENOVO] JVM arguments [-XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -XX:+HeapDumpOnOutOfMemoryError, -Xmx2048m, -Xms2048m, -Delasticsearch, -Des.path.home=C:\Program Files\Elastic\Elasticsearch\6.1.2, -Des.path.conf=C:\ProgramData\Elastic\Elasticsearch\config]
[2018-02-08T10:57:54,445][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [aggs-matrix-stats]
[2018-02-08T10:57:54,445][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [analysis-common]
[2018-02-08T10:57:54,445][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [ingest-common]
[2018-02-08T10:57:54,445][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [lang-expression]
[2018-02-08T10:57:54,445][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [lang-mustache]
[2018-02-08T10:57:54,445][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [lang-painless]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [mapper-extras]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [parent-join]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [percolator]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [reindex]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [repository-url]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [transport-netty4]
[2018-02-08T10:57:54,446][INFO ][o.e.p.PluginsService     ] [LENOVO] loaded module [tribe]
[2018-02-08T10:57:54,447][INFO ][o.e.p.PluginsService     ] [LENOVO] no plugins loaded
[2018-02-08T10:57:55,044][ERROR][o.e.b.Bootstrap          ] Exception
java.lang.IllegalArgumentException: unknown secure setting [bootstrap.password] please check that any required plugins are installed, or check the breaking changes documentation for removed settings
	at org.elasticsearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:312) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.common.settings.AbstractScopedSettings.validate(AbstractScopedSettings.java:276) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.common.settings.SettingsModule.<init>(SettingsModule.java:135) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.node.Node.<init>(Node.java:330) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.node.Node.<init>(Node.java:245) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:212) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:212) ~[elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:322) [elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:121) [elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:112) [elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86) [elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124) [elasticsearch-cli-6.1.2.jar:6.1.2]
	at org.elasticsearch.cli.Command.main(Command.java:90) [elasticsearch-cli-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:92) [elasticsearch-6.1.2.jar:6.1.2]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:85) [elasticsearch-6.1.2.jar:6.1.2]
[2018-02-08T10:57:55,050][WARN ][o.e.b.ElasticsearchUncaughtExceptionHandler] [LENOVO] uncaught exception in thread [main]
```
**The installer however still succeeds in this case**

To be clear, building 6.1.1 and 6.1.2 from the 6.1 branch exhibits the same behaviour, that of a successful upgrade even though the service started then stopped, possibly because of the [process exit handler(s)](https://github.com/elastic/windows-installers/blob/a9c71968a56f946116830f4abe74f4a9733066e4/src/ProcessHosts/Elastic.ProcessHosts/Process/ObservableProcess.cs#L40-L48) signalling completion (it's not clear to me which `process.Exited` event handler is calling `observer.OnCompleted();` here; this needs investigation).

To my original comment

>I totally see the merits to taking the assume started approach, which implies a philosophy of best effort to starting the service.

it could be argued that the service _is_ started as part of installation in this case, and that its subsequent _near immediate_ stopping because of failed bootstrap checks is not a concern of the installer and installation process and hence not grounds to fail the installation, which aligns with the philosophy of assume started. _Conversely_, it could be argued that the service immedately stopping because of bootstrap checks is grounds to fail the installation process. This is an installer design discussion, and I have no strong opinions of either approach 😄 If I look at it from a users perspective, which approach would be a better experience?

Reply:
We settled offline that this is currently the best way forward but lets re discuss when you get back @russcam 👍